lib/TPP/Transforms/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,179 @@
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+static ArrayAttr getBrgemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::BRGEMM_INS,
+                                   memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto *output = brgemmOp.getOperand(3).getDefiningOp();
+    if (!output)
+      return failure();
+
+    // First, match the required fused ops
+    auto result = xsmm::utils::getFusedBrgemmSequenceFromProducer(output);
+    if (failed(result))
+      return failure();
+    auto fusedMatch = *result;
+    // TODO: Support BRGEMM + BINARY && BRGEMM + UNARY patterns
+    if (!fusedMatch.binaryOp || !fusedMatch.unaryOp)
+      return failure();
+    // Validate broadcast flags
+    auto unaryFlags =
+        xsmm::utils::getUnaryFlags(fusedMatch.unaryOp.getOperand(0).getType(),
+                                   fusedMatch.unaryOp.getOperand(2).getType());
+    if (unaryFlags != mlir::xsmm::UnaryFlags::BCAST_SCALAR &&
+        unaryFlags != mlir::xsmm::UnaryFlags::NONE)
+      return failure();
+
+    // TODO: Support more than just COL_0 BCAST
+    auto binaryFlags =
+        xsmm::utils::getBinaryFlags(fusedMatch.binaryOp.getOperand(1).getType(),
+                                    fusedMatch.binaryOp.getOperand(3).getType(),
+                                    mlir::xsmm::utils::OperandPos::LHS);
+    if (binaryFlags != mlir::xsmm::BinaryFlags::BCAST_COL_IN_0)
+      return failure();
+
+    // Now, replace the ops with a fused BRGEMM
+    auto dtype =
+        xsmm::utils::getDataType(rewriter, brgemmOp.getOperand(1).getType());
+    IntegerType integer64 = IntegerType::get(rewriter.getContext(), 64);
+
+    Location loc = brgemmOp.getLoc();
+    auto dims = getSizesAndLeadingDimForBrgemmOp(rewriter, brgemmOp);
+    auto memrefB = brgemmOp.getOperand(2);
+    int64_t batchSize = memrefB.getType().cast<ShapedType>().getShape()[0];
+
+    OpBuilder::InsertionGuard guard(rewriter);
```

Comment:
why are you changing the insertion point?

Reply:
Because the matcher starts at the BRGEMM location but in case there's an ADD, the subview for the bias is created _after_ the BRGEMM. If we insert the fused BRGEMM _before_ the bias subview, we create an IR problem.

This insertion point is to make sure we place the fused BRGEMM at the binary op level (because the unary doesn't need any extra subviews).